### PR TITLE
Build SOF with Zephyr on i.MX

### PR DESCRIPTION
**Pull request by iuliana-prodan**
Add support for i.MX to build Sound Open Firmware with Zephyr OS.

Please check each commit message.

With these commits I have a successful build. 
I obtain the zephyr.ri image which boots the DSP and loads the topology.

```
root@imx8qmmek:~# ./loadSof.sh 
[  166.405965] sof-audio-of 556e8000.dsp: DT DSP detected
[  166.416971] imx-dsp imx-dsp: NXP i.MX DSP IPC initialized
[  166.430784] sof-audio-of 556e8000.dsp: warning: unknown sof_ext_man header type 5 size 0x20
[  166.439213] sof-audio-of 556e8000.dsp: warning: unknown sof_ext_man header type 3 size 0x30
[  166.447885] sof-audio-of 556e8000.dsp: Firmware info: version 1:8:0-cf919
[  166.454703] sof-audio-of 556e8000.dsp: Firmware: ABI 3:19:0 Kernel ABI 3:17:0
[  166.461862] sof-audio-of 556e8000.dsp: warn: FW ABI is more recent than kernel
[  166.469453] sof-audio-of 556e8000.dsp: Firmware info: version 1:8:0-cf919
[  166.476264] sof-audio-of 556e8000.dsp: Firmware: ABI 3:19:0 Kernel ABI 3:17:0
[  166.483406] sof-audio-of 556e8000.dsp: warn: FW ABI is more recent than kernel
[  166.506546] sof-audio-of 556e8000.dsp: Topology: ABI 3:18:1 Kernel ABI 3:17:0
[  166.515821] sof-audio-of 556e8000.dsp: warn: topology ABI is more recent than kernel
[  166.526795] sof-audio-of 556e8000.dsp: tplg: config SAI1 fmt 0x1 mclk 12288000 width 32 slots 2 mclk id 0
[  166.537294] sof-audio-of 556e8000.dsp: ASoC: Parent card not yet available, widget card binding deferred
root@imx8qmmek:~# 
root@imx8qmmek:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: btscoaudio [bt-sco-audio], device 0: 59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0 [59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sofwm8960audio [sof-wm8960-audio], device 0: Port0 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
root@imx8qmmek:~# arecord -l
**** List of CAPTURE Hardware Devices ****
card 0: btscoaudio [bt-sco-audio], device 0: 59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0 [59040000.sai-bt-sco-pcm-wb bt-sco-pcm-wb-0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: sofwm8960audio [sof-wm8960-audio], device 0: Port0 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
Pull requests were also opened on:
- zephyr repo - see [#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467) 
- xtensa_hal - see [#12](https://github.com/zephyrproject-rtos/hal_xtensa/pull/12)

### CODE REVIEW

**Review by lgirdwood — COMMENTED** [review]
Can we also add the imx Zephyr targets into the build script used by CI.

**Comment by lgirdwood** [comment]
> Sure, I can do that, but shouldn't we wait for the Zephy PR be merged - [#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467)?

Yes, sorry I was meaning to make the script build imx Zephyr as part of this PR.

**Comment by iuliana-prodan** [comment]
> > Sure, I can do that, but shouldn't we wait for the Zephy PR be merged - [#35467](https://github.com/zephyrproject-rtos/zephyr/pull/35467)?
> 
> Yes, sorry I was meaning to make the script build imx Zephyr as part of this PR.

Sure, I'll do that.
I also have another suggestion from @lyakh to fix some warnings on trace.h.

File: zephyr/CMakeLists.txt
Comment (dbaluta) [inline zephyr/CMakeLists.txt:350]: @iuliana-prodan we need to be aware that we have two platforms (imx8 and imx8x) although very similar. Not sure what PLATFORM stands for here.

File: zephyr/CMakeLists.txt
Comment (dbaluta) [inline zephyr/CMakeLists.txt:374]: @iuliana-prodan I don't see these includes for imx8 above. Is this ok?

File: zephyr/wrapper.c
Comment (dbaluta) [inline zephyr/wrapper.c:205]: I would leave CONFIG_LIBRARY as the last CONFIG in this line as it special. Not a real platform.

File: zephyr/wrapper.c
Comment (dbaluta) [inline zephyr/wrapper.c:208]: No need for this new line here.

File: zephyr/wrapper.c
Comment (dbaluta) [inline zephyr/wrapper.c:347]: My understanding here is that we want an implementation for platform_timer_get() that doesn't use XTOS API at all but it uses APIs from Zephyr. 

@lgirdwood is this true?

File: zephyr/CMakeLists.txt
Comment (iuliana-prodan) [reply zephyr/CMakeLists.txt:350]: This part here includes all the files from sof/src/platform/imx8/ which are common for both imx8 and imx8x.
I've set PLATFORM as imx8 since this is the only one I've tested and focused my work on this.

When I'll check on imx8x will see if I need to modify this or not. Right now, I'm not sure.

File: zephyr/CMakeLists.txt
Comment (iuliana-prodan) [reply zephyr/CMakeLists.txt:374]: These were not needed for imx8 and I've removed them.
Probably they will not be needed for imx8m either. 
I'll remove them, for now, and see what happens when testing on i.MX8MP.

File: zephyr/wrapper.c
Comment (iuliana-prodan) [reply zephyr/wrapper.c:347]: Seeing the "TODO: align with 64bit Zephyr timers when they are upstream." from [here](https://github.com/thesofproject/sof/blob/main/zephyr/wrapper.c#L276), for now, I've kept the Xtensa timer.

From my understanding Intel does the same with CAVS platforms supported in Zephyr - see [here](https://github.com/thesofproject/sof/blob/main/zephyr/wrapper.c#L308)

But, I'll let @lgirdwood comment also.

File: zephyr/CMakeLists.txt
Comment (paulstelian97) [reply zephyr/CMakeLists.txt:350]: imx8 and imx8x have minor differences in the kconfig, really. They are extremely similar but not the same. Basically all the code is shared but addresses of hardware and shit differ.

File: zephyr/wrapper.c
Comment (lgirdwood) [reply zephyr/wrapper.c:347]: yes, we are moving to use Zephyr native APIs through out the codebase. @kv2019i @marc-hb and @lyakh are working on this. The intention is to gradually remove code from zephyr/wrapper.c and have an xtos/wrapper.c (and eventually deprecate xtos).

File: zephyr/CMakeLists.txt
Comment (iuliana-prodan) [reply zephyr/CMakeLists.txt:372]: @dbaluta I've removed them.
I'll check if those includes are needed when testing imx8m.

File: zephyr/wrapper.c
Comment (iuliana-prodan) [reply zephyr/wrapper.c]: Done

File: zephyr/wrapper.c
Comment (iuliana-prodan) [reply zephyr/wrapper.c:207]: I removed it.

**Review by lgirdwood — APPROVED** [review]
LGTM, but I wont merge until NXP folks have aligned or merge it themselves.

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
LGTM, thanks Iulia!

File: zephyr/wrapper.c
Comment (marc-hb) [inline zephyr/wrapper.c:155]: This line "reliably" crashes my Up Squared board on the second aplay command every time, see partial revert PR #4472

git bisect found this commit and then it was easy to identify this line because it seems to be the only one not IMX-specific.

Cc: @lyakh , @lgirdwood , @andyross

Comment (iuliana-prodan) [reply zephyr/wrapper.c:155]: I've added a pull request to fix this, for both i.MX and Up Squared: #4477 
The PR #4472 breaks i.MX.

**Pull request merged**
